From e9709cefd9dcdc8c96df5a135c8f039a037a1a72 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 7 Apr 2021 08:25:57 +0800
Subject: [PATCH 47/93] config-parser: Support loading multiple configs

Try loading .ini configs under "<config>.d/".

Tested with:
/etc/xdg/weston/weston.ini.d/99-pixman.ini
[core]
use-pixman=true

And strip leading spaces.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 shared/config-parser.c | 129 +++++++++++++++++++++++++++++++++++------
 1 file changed, 110 insertions(+), 19 deletions(-)

diff --git a/shared/config-parser.c b/shared/config-parser.c
index 30779ae..8ad9ceb 100644
--- a/shared/config-parser.c
+++ b/shared/config-parser.c
@@ -31,6 +31,7 @@
 #include <stdlib.h>
 #include <assert.h>
 #include <ctype.h>
+#include <dirent.h>
 #include <limits.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -70,6 +71,13 @@ open_config_file(struct weston_config *c, const char *name)
 	const char *p, *next;
 	int fd;
 
+	if (!c) {
+		if (name[0] != '/')
+			return -1;
+
+		return open(name, O_RDONLY | O_CLOEXEC);
+	}
+
 	if (name[0] == '/') {
 		snprintf(c->path, sizeof c->path, "%s", name);
 		return open(name, O_RDONLY | O_CLOEXEC);
@@ -337,6 +345,15 @@ config_add_section(struct weston_config *config, const char *name)
 {
 	struct weston_config_section *section;
 
+	/* squash single sessions */
+	if (strcmp(name, "launcher") && strcmp(name, "ivi-launcher") &&
+	    strcmp(name, "output") && strcmp(name, "remote-output") &&
+	    strcmp(name, "pipewire-output")) {
+		section = weston_config_get_section(config, name, NULL, NULL);
+		if (section)
+			return section;
+	}
+
 	section = zalloc(sizeof *section);
 	if (section == NULL)
 		return NULL;
@@ -359,6 +376,24 @@ section_add_entry(struct weston_config_section *section,
 {
 	struct weston_config_entry *entry;
 
+	/* hack for removing entry */
+	if (key[0] == '-') {
+		key ++;
+		value = NULL;
+	}
+
+	/* drop old entry */
+	entry = config_section_get_entry(section, key);
+	if (entry) {
+		wl_list_remove(&entry->link);
+		free(entry->key);
+		free(entry->value);
+		free(entry);
+	}
+
+	if (!value || value[0] == '\0')
+		return NULL;
+
 	entry = zalloc(sizeof *entry);
 	if (entry == NULL)
 		return NULL;
@@ -388,8 +423,6 @@ weston_config_parse_internal(struct weston_config *config, FILE *fp)
 	char line[512], *p;
 	int i;
 
-	wl_list_init(&config->section_list);
-
 	while (fgets(line, sizeof line, fp)) {
 		switch (line[0]) {
 		case '#':
@@ -438,6 +471,7 @@ weston_config_parse_fp(FILE *file)
 	if (config == NULL)
 		return NULL;
 
+	wl_list_init(&config->section_list);
 	if (!weston_config_parse_internal(config, file)) {
 		weston_config_destroy(config);
 		return NULL;
@@ -446,48 +480,105 @@ weston_config_parse_fp(FILE *file)
 	return config;
 }
 
-WL_EXPORT struct weston_config *
-weston_config_parse(const char *name)
+static FILE *
+weston_open_config_file(struct weston_config *config, const char *name)
 {
 	FILE *fp;
 	struct stat filestat;
-	struct weston_config *config;
 	int fd;
-	bool ret;
-
-	config = zalloc(sizeof *config);
-	if (config == NULL)
-		return NULL;
 
 	fd = open_config_file(config, name);
-	if (fd == -1) {
-		free(config);
+	if (fd == -1)
 		return NULL;
-	}
 
 	if (fstat(fd, &filestat) < 0 ||
 	    !S_ISREG(filestat.st_mode)) {
 		close(fd);
-		free(config);
 		return NULL;
 	}
 
 	fp = fdopen(fd, "r");
 	if (fp == NULL) {
 		close(fd);
-		free(config);
 		return NULL;
 	}
 
-	ret = weston_config_parse_internal(config, fp);
+	return fp;
+}
 
-	fclose(fp);
+static int
+accept_config_file(const struct dirent *entry)
+{
+	const char *suffix = ".ini";
+	char *end = strstr(entry->d_name, suffix);
+	return end && end[strlen(suffix)] == '\0';
+}
 
-	if (!ret) {
-		weston_config_destroy(config);
+WL_EXPORT struct weston_config *
+weston_config_parse(const char *name)
+{
+	FILE *fp;
+	struct weston_config *config;
+	struct stat st;
+	struct dirent **namelist;
+	char path[sizeof(config->path) + 2];
+	bool ret;
+	int n;
+
+	config = zalloc(sizeof *config);
+	if (config == NULL)
 		return NULL;
+
+	wl_list_init(&config->section_list);
+
+	fp = weston_open_config_file(config, name);
+	if (fp) {
+		ret = weston_config_parse_internal(config, fp);
+
+		fclose(fp);
+
+		if (!ret) {
+			fprintf(stderr, "failed to parse %s\n", config->path);
+			free(config);
+			return NULL;
+		}
+	}
+
+	strcpy(path, config->path);
+	strcat(path, ".d");
+	if (stat(path, &st) < 0 || !S_ISDIR(st.st_mode))
+		return config;
+
+	n = scandir(path, &namelist, accept_config_file, alphasort);
+	if (n < 0)
+		return config;
+
+	while (n--) {
+		char *file = namelist[n]->d_name;
+		char *sep = "/";
+		char fpath[strlen(path)+strlen(sep)+strlen(file) + 1];
+		strcpy(fpath, path);
+		strcat(fpath, sep);
+		strcat(fpath, file);
+		free(namelist[n]);
+
+		fp = weston_open_config_file(NULL, fpath);
+		if (!fp)
+			continue;
+
+		ret = weston_config_parse_internal(config, fp);
+
+		fclose(fp);
+
+		if (!ret) {
+			fprintf(stderr, "failed to parse '%s'\n", fpath);
+			free(namelist);
+			free(config);
+			return NULL;
+		}
 	}
 
+	free(namelist);
 	return config;
 }
 
-- 
2.20.1

